File: apps/desktop/src-tauri/src/downloads/engine.rs
//! Orchestrateur des transferts : 2 téléchargements simultanés max, FIFO,
//! reprise au démarrage (pauses SYSTÈME uniquement), pause/reprise/annulation.
//! Les identifiants de connexion (serveur + token) vivent EN MÉMOIRE seulement,
//! fournis par le front à chaque session — jamais écrits en base.

use super::{db, fsops, meta, queue, store, transfer};
use serde::Serialize;
use std::collections::HashMap;
use std::sync::{Arc, Mutex, MutexGuard, OnceLock};
use tauri::{AppHandle, Emitter};

pub const MAX_PARALLEL: usize = 2;
pub const EVENT_PROGRESS: &str = "downloads://progress";
pub const EVENT_CHANGED: &str = "downloads://changed";

#[derive(Clone)]
pub struct Creds {
    pub server_url: String,
    pub token: String,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
struct ProgressPayload {
    file_id: i64,
    bytes_done: i64,
    expected_size: Option<i64>,
}

pub struct Engine {
    inner: Arc<EngineInner>,
}

impl Clone for Engine {
    fn clone(&self) -> Self {
        Self { inner: self.inner.clone() }
    }
}

struct EngineInner {
    app: OnceLock<AppHandle>,
    creds: Mutex<Option<Creds>>,
    active: Mutex<HashMap<i64, Arc<transfer::TransferFlags>>>,
}

fn lock<T>(mutex: &Mutex<T>) -> MutexGuard<'_, T> {
    match mutex.lock() {
        Ok(guard) => guard,
        Err(poisoned) => poisoned.into_inner(),
    }
}

pub fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

impl Default for Engine {
    fn default() -> Self {
        Self::new()
    }
}

impl Engine {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(EngineInner {
                app: OnceLock::new(),
                creds: Mutex::new(None),
                active: Mutex::new(HashMap::new()),
            }),
        }
    }

    fn app(&self) -> Option<AppHandle> {
        self.inner.app.get().cloned()
    }

    fn emit_changed(&self) {
        if let Some(app) = self.app() {
            let _ = app.emit(EVENT_CHANGED, ());
        }
    }

    pub fn is_active(&self, file_id: i64) -> bool {
        lock(&self.inner.active).contains_key(&file_id)
    }

    /// Notification « quelque chose a changé » vers l'UI (listes à invalider).
    pub fn notify_changed(&self) {
        self.emit_changed();
    }

    /// Démarrage/reconnexion : pose les credentials, normalise la file
    /// (interrompus + pauses système → queued) et relance.
    pub fn start(&self, app: &AppHandle, creds: Creds) -> Result<(), String> {
        let _ = self.inner.app.set(app.clone());
        *lock(&self.inner.creds) = Some(creds);
        let conn = db::open(&db::db_path(app)?)?;
        queue::normalize_on_engine_start(&conn, now_ms())?;
        drop(conn);
        self.emit_changed();
        self.pump();
        Ok(())
    }

    /// Rafraîchit les credentials sans re-normaliser (enqueue courant).
    pub fn set_creds(&self, app: &AppHandle, creds: Creds) {
        let _ = self.inner.app.set(app.clone());
        *lock(&self.inner.creds) = Some(creds);
    }

    /// Lance des transferts tant qu'il y a des slots ET des fichiers en file.
    pub fn pump(&self) {
        let Some(app) = self.app() else { return };
        let Some(creds) = lock(&self.inner.creds).clone() else { return };
        loop {
            let Ok(db_path) = db::db_path(&app) else { return };
            let Ok(conn) = db::open(&db_path) else { return };
            let mut active = lock(&self.inner.active);
            if active.len() >= MAX_PARALLEL {
                return;
            }
            let Ok(Some(file)) = queue::next_queued(&conn) else { return };
            if queue::set_status(&conn, file.id, "downloading", None, now_ms()).is_err() {
                return;
            }
            let flags = Arc::new(transfer::TransferFlags::new());
            active.insert(file.id, flags.clone());
            drop(active);
            drop(conn);
            self.spawn_worker(app.clone(), creds.clone(), file, flags);
        }
    }

    fn spawn_worker(
        &self,
        app: AppHandle,
        creds: Creds,
        file: store::FileRow,
        flags: Arc<transfer::TransferFlags>,
    ) {
        let engine = self.clone();
        std::thread::spawn(move || {
            let end = run_worker(&app, &creds, &file, &flags);
            engine.finish(&app, file.id, end);
        });
    }

    fn finish(&self, app: &AppHandle, file_id: i64, end: transfer::TransferEnd) {
        lock(&self.inner.active).remove(&file_id);
        if let Ok(path) = db::db_path(app) {
            if let Ok(conn) = db::open(&path) {
                let now = now_ms();
                use transfer::TransferEnd as E;
                let _ = match end {
                    E::Complete { final_size } => {
                        let _ = queue::set_bytes_done(&conn, file_id, final_size, now);
                        queue::set_status(&conn, file_id, "complete", None, now)
                    }
                    E::Paused { bytes_done } => {
                        let _ = queue::set_bytes_done(&conn, file_id, bytes_done, now);
                        queue::set_status(&conn, file_id, "paused", None, now)
                    }
                    E::Canceled => {
                        let _ = queue::set_bytes_done(&conn, file_id, 0, now);
                        queue::set_status(&conn, file_id, "canceled", None, now)
                    }
                    E::Failed { code, bytes_done } => {
                        let _ = queue::set_bytes_done(&conn, file_id, bytes_done, now);
                        if code == "network" {
                            // Coupure réseau = pause SYSTÈME → auto-reprise.
                            let _ = queue::set_paused_by_user(&conn, file_id, false);
                            queue::set_status(&conn, file_id, "paused", None, now)
                        } else {
                            queue::set_status(&conn, file_id, "error", Some(code), now)
                        }
                    }
                };
            }
        }
        self.emit_changed();
        self.pump();
    }

    pub fn pause(&self, app: &AppHandle, file_id: i64) -> Result<(), String> {
        let conn = db::open(&db::db_path(app)?)?;
        queue::set_paused_by_user(&conn, file_id, true)?;
        if let Some(flags) = lock(&self.inner.active).get(&file_id) {
            flags.pause.store(true, std::sync::atomic::Ordering::Relaxed);
        } else if let Some(file) = queue::get_file(&conn, file_id)? {
            if file.status == "queued" {
                queue::set_status(&conn, file_id, "paused", None, now_ms())?;
            }
        }
        self.emit_changed();
        Ok(())
    }

    pub fn resume(&self, app: &AppHandle, file_id: i64) -> Result<(), String> {
        let conn = db::open(&db::db_path(app)?)?;
        if let Some(file) = queue::get_file(&conn, file_id)? {
            if file.status == "paused" || file.status == "error" {
                queue::set_paused_by_user(&conn, file_id, false)?;
                queue::set_status(&conn, file_id, "queued", None, now_ms())?;
            }
        }
        drop(conn);
        self.emit_changed();
        self.pump();
        Ok(())
    }

    pub fn cancel(&self, app: &AppHandle, file_id: i64) -> Result<(), String> {
        if let Some(flags) = lock(&self.inner.active).get(&file_id) {
            flags.cancel.store(true, std::sync::atomic::Ordering::Relaxed);
            return Ok(()); // le worker nettoie le .part et pose le statut
        }
        let conn = db::open(&db::db_path(app)?)?;
        if let Some(file) = queue::get_file(&conn, file_id)? {
            let root = fsops::resolve_root(app)?;
            fsops::remove_media_file(&root, &file.rel_path)?;
            queue::set_bytes_done(&conn, file_id, 0, now_ms())?;
            queue::set_status(&conn, file_id, "canceled", None, now_ms())?;
        }
        self.emit_changed();
        Ok(())
    }

    /// Attend la fin effective d'un worker (suppression de claim sûre).
    pub fn wait_not_active(&self, file_id: i64, timeout_ms: u64) {
        let deadline = std::time::Instant::now() + std::time::Duration::from_millis(timeout_ms);
        while self.is_active(file_id) && std::time::Instant::now() < deadline {
            std::thread::sleep(std::time::Duration::from_millis(50));
        }
    }
}

/// Corps du worker : snapshot méta best-effort puis transfert streaming.
fn run_worker(
    app: &AppHandle,
    creds: &Creds,
    file: &store::FileRow,
    flags: &transfer::TransferFlags,
) -> transfer::TransferEnd {
    let Ok(root) = fsops::resolve_root(app) else {
        return transfer::TransferEnd::Failed { code: "io", bytes_done: file.bytes_done };
    };
    let Ok(db_path) = db::db_path(app) else {
        return transfer::TransferEnd::Failed { code: "io", bytes_done: file.bytes_done };
    };
    let Ok(conn) = db::open(&db_path) else {
        return transfer::TransferEnd::Failed { code: "io", bytes_done: file.bytes_done };
    };

    let agent = ureq::AgentBuilder::new()
        .timeout(std::time::Duration::from_secs(20))
        .build();
    if !meta::snapshot_exists(&root, &file.item_id) {
        if let Ok(Some(spec)) = meta::get_spec(&conn, &file.item_id) {
            let _ = meta::snapshot(&agent, &creds.server_url, &creds.token, &root, &conn, &spec);
        }
    }
    // Sous-titres texte en side-cars (les deux variantes) — best-effort.
    if let Some(json) = &file.subtitles_json {
        let specs = super::subs::parse_specs(json);
        if !specs.is_empty() {
            let _ = super::subs::fetch_all(
                &agent, &creds.server_url, &creds.token, &root,
                &file.item_id, &file.media_source_id, &specs,
            );
        }
    }

    let Ok(final_path) = fsops::safe_join(&root, &file.rel_path) else {
        return transfer::TransferEnd::Failed { code: "io", bytes_done: 0 };
    };
    let url = match file.variant.as_str() {
        "original" => format!(
            "{}/api/downloads/original/{}?mediaSourceId={}",
            creds.server_url, file.item_id, file.media_source_id
        ),
        _ => {
            let mut light_url = format!(
                "{}/api/downloads/light/{}?mediaSourceId={}&preset={}",
                creds.server_url,
                file.item_id,
                file.media_source_id,
                file.preset.as_deref().unwrap_or("p720")
            );
            if let Some(audio) = file.audio_stream_index {
                light_url.push_str(&format!("&audioStreamIndex={audio}"));
            }
            if let Some(burn) = file.burn_subtitle_index {
                light_url.push_str(&format!("&burnSubtitleIndex={burn}"));
            }
            light_url
        }
    };
    let job = transfer::TransferJob {
        url,
        token: creds.token.clone(),
        final_path,
        variant: file.variant.clone(),
        expected_size: file.expected_size,
        server_url: creds.server_url.clone(),
    };
    let app_for_events = app.clone();
    let file_id = file.id;
    let expected = file.expected_size;
    transfer::run(&job, flags, &move |bytes| {
        let _ = queue::set_bytes_done(&conn, file_id, bytes, now_ms());
        let _ = app_for_events.emit(
            EVENT_PROGRESS,
            ProgressPayload { file_id, bytes_done: bytes, expected_size: expected },
        );
    })
}

File: packages/offline-core/src/core/engine.ts
/**
 * Orchestrateur des transferts : un seul à la fois, FIFO, reprise au
 * démarrage, pause, reprise, annulation.
 *
 * Ce qui suit la fin d'un transfert — sa traduction en statut, la finalisation
 * du fichier Allégé — vit dans `engineEnd.ts` ; les dépendances de plateforme
 * dans `engineDeps.ts`.
 *
 * Les identifiants de connexion vivent EN MÉMOIRE seulement, fournis par la
 * page à chaque session — jamais écrits en base.
 *
 * # La différence avec le Rust, et pourquoi elle simplifie
 *
 * `engine.rs` lançait un THREAD par transfert, avec une connexion SQLite
 * chacun, un `Mutex` sur la table des actifs et des `AtomicBool` pour les
 * bascules. Ici tout vit sur la boucle d'évènements : les transferts sont des
 * fonctions asynchrones, la table des actifs est une `Map` ordinaire, et les
 * bascules des booléens. Il n'y a aucune concurrence à protéger — seulement de
 * l'entrelacement, qui ne se produit qu'aux `await`.
 */

import type { EngineDeps } from "./engineDeps";
import { applyEnd, mediaAwaitingFinalize, runFinalize } from "./engineEnd";
import { cancelFile, pauseFile, resumeFile } from "./engineGestures";
import { RETRY_DELAYS_MS, RetryScheduler } from "./retry";
import {
  countQueued,
  getFile,
  nextQueued,
  normalizeOnEngineStart,
  requeueSystemPauses,
  setBytesDone,
  setExpectedSize,
  setStatus,
  suspendQueued,
} from "./queue";
import type { FileRow } from "./store";
import { TransferFlags, type TransferEnd } from "./transfer";
import { runWorker, type Creds } from "./worker";

export type { EngineDeps, FinalizeVerdict } from "./engineDeps";

/**
 * Un seul transfert à la fois. Deux se disputaient la bande passante, le disque
 * et — sur le téléphone — le processeur de la finalisation MP4 : deux barres
 * qui avancent lentement plutôt qu'une qui aboutit. La file reste FIFO.
 *
 * La plateforme peut en demander plus le temps d'une suspension système
 * (cf. `EngineDeps.parallelLimit`) : là, ce qui n'est pas déjà parti ne
 * partira pas du tout.
 */
export const MAX_PARALLEL = 1;

/** Garde-fou : au-delà, on sature la connexion sans rien terminer plus vite. */
const PARALLEL_CEILING = 4;

export class DownloadEngine {
  private creds: Creds | null = null;
  private readonly active = new Map<number, TransferFlags>();
  private busy = false;
  /** Entre `suspendForSystem` et `resumeSystemPauses` : les conditions manquent. */
  private systemSuspended = false;
  private readonly retries: RetryScheduler;

  constructor(private readonly deps: EngineDeps) {
    this.retries = new RetryScheduler(deps.db, deps.now, () => this.sweepRetries());
  }

  isActive(fileId: number): boolean {
    return this.active.has(fileId);
  }

  /**
   * Transferts en cours ET en attente — ce qu'une fermeture interromprait.
   *
   * Sans identifiants, le moteur ne peut rien tirer : une file qui attend une
   * reconnexion n'est pas « en cours », et le dire ferait poser une question à
   * l'utilisateur pour un transfert qui, de toute façon, ne bouge pas.
   */
  pending(): number {
    if (this.creds === null) return 0;
    return this.active.size + countQueued(this.deps.db);
  }

  /** « Quelque chose a changé » : l'interface invalide ses listes. */
  notifyChanged(): void {
    this.deps.emit("downloads://changed", undefined);
  }

  /**
   * Démarrage ou reconnexion : pose les identifiants, normalise la file
   * (transferts interrompus et pauses système → `queued`), relance.
   */
  start(creds: Creds): void {
    this.creds = creds;
    this.systemSuspended = false;
    // Moteur vivant (reconnexion) : les transferts qui tournent ne sont pas
    // « interrompus » — les remettre en file les lancerait deux fois sur le
    // même `.part`. Seules les pauses système sont alors rattrapées.
    if (this.active.size === 0) normalizeOnEngineStart(this.deps.db, this.deps.now());
    else requeueSystemPauses(this.deps.db, this.deps.now());
    this.notifyChanged();
    this.pump();
    this.deps.onStarted?.(creds);
  }

  /** Rafraîchit les identifiants sans re-normaliser (mise en file courante). */
  setCreds(creds: Creds): void {
    this.creds = creds;
  }

  /**
   * Lance des transferts tant qu'il y a des places ET des fichiers en file.
   *
   * PASSAGE OBLIGÉ de toute variation d'activité : `start`, la mise en file,
   * `resume` et la fin d'un transfert (`finish`) appellent tous `pump`. La
   * bascule occupé/inoccupé est donc calculée ici, une fois, plutôt que
   * dispersée dans quatre appelants qui finiraient par en oublier un.
   */
  pump(): void {
    if (this.deps.canTransfer?.() === false) {
      // Une seule garde, ici : la mise en file, la reprise explicite, le
      // démarrage et le retour au premier plan passent tous par `pump`.
      if (suspendQueued(this.deps.db, this.deps.now()) > 0) this.notifyChanged();
    } else {
      this.startWhatCanRun();
    }
    this.updateActivity();
    this.retries.arm();
  }

  /**
   * Remet en file ce dont l'échéance de relance est atteinte.
   *
   * Public parce que le minuteur ne suffit pas sur le téléphone : le système
   * gèle les minuteurs en arrière-plan, et le retour au premier plan doit
   * rattraper les échéances passées entre-temps.
   */
  sweepRetries(): void {
    if (this.retries.sweep()) {
      this.notifyChanged();
      this.pump();
    } else {
      this.retries.arm();
    }
  }

  /** Places ouvertes maintenant, bornées : la plateforme demande, elle n'impose pas. */
  private parallelLimit(): number {
    const asked = this.deps.parallelLimit?.() ?? MAX_PARALLEL;
    return Math.min(Math.max(1, Math.trunc(asked)), PARALLEL_CEILING);
  }

  private startWhatCanRun(): void {
    const creds = this.creds;
    if (creds === null) return;
    const limit = this.parallelLimit();
    while (this.active.size < limit) {
      const file = nextQueued(this.deps.db);
      if (file === null) return;
      // Le statut passe à `downloading` AVANT le premier `await` : sans ça, le
      // tour de boucle suivant reprendrait le même fichier.
      setStatus(this.deps.db, file.id, "downloading", null, this.deps.now());
      const flags = new TransferFlags();
      this.active.set(file.id, flags);
      void this.work(creds, file.id, flags);
    }
  }

  /** Ne notifie qu'aux transitions — voir `EngineDeps.onBusy`. */
  private updateActivity(): void {
    const busy = this.active.size > 0;
    if (busy === this.busy) return;
    this.busy = busy;
    this.deps.onBusy?.(busy);
  }

  private async work(creds: Creds, fileId: number, flags: TransferFlags): Promise<void> {
    const file = getFile(this.deps.db, fileId);
    let end: TransferEnd;
    const received =
      file === null || this.deps.finalizeMedia === undefined
        ? null
        : mediaAwaitingFinalize(this.deps.db, this.deps.volume(), file, this.deps.now());
    if (file === null) {
      end = { kind: "failed", code: "io", bytesDone: 0 };
    } else if (received !== null) {
      // Média entièrement reçu, seul le remux avait échoué : on le retente, et
      // rien ne repart sur le réseau.
      end = await this.finalize(file, received);
    } else {
      // Mutable : le pilote peut annoncer le total en cours de route, et une
      // valeur gelée ici laisserait la barre sans repère jusqu'à la fin.
      let expected = file.expectedSize;
      try {
        end = await runWorker(
          {
            db: this.deps.db,
            volume: this.deps.volume(),
            driver: this.deps.driver,
            fetchBytes: this.deps.makeFetcher(creds.token),
            onProgress: (id, bytes) => this.progress(id, bytes, expected),
            onExpected: (id, totalBytes) => {
              expected = totalBytes;
              setExpectedSize(this.deps.db, id, totalBytes, this.deps.now());
            },
            onUnexpected: this.deps.onUnexpected,
            now: this.deps.now,
          },
          creds,
          file,
          flags,
          this.deps.now(),
        );
      } catch (error) {
        // Une exception levée PARCE QU'ON VIENT D'INTERROMPRE n'est pas une
        // panne : le pilote a été coupé net, c'est tout ce qu'elle dit. La
        // traduire en « imprévu » armait une relance à cinq secondes, qui
        // faisait repartir ce que l'utilisateur venait d'arrêter — depuis la
        // notification Android comme depuis la feuille d'actions.
        if (flags.cancel) end = { kind: "canceled" };
        else if (flags.pause) end = { kind: "paused", bytesDone: file.bytesDone };
        else {
          // Un défaut inattendu ne doit pas laisser le fichier en `downloading`
          // pour l'éternité — il resterait invisible jusqu'au prochain
          // démarrage. La trace est tout ce qui restera pour comprendre : le
          // statut, lui, ne dit rien de plus que « imprévu ».
          this.deps.onUnexpected?.("engine.work", error);
          end = { kind: "failed", code: "unexpected", bytesDone: file.bytesDone };
        }
      }
      if (end.kind === "complete" && file.variant === "light" && this.deps.finalizeMedia !== undefined) {
        end = await this.finalize(file, end.finalSize);
      }
    }
    this.finish(fileId, end);
  }

  private finalize(file: FileRow, finalSize: number): Promise<TransferEnd> {
    return runFinalize(
      this.deps.db,
      this.deps.volume(),
      file.id,
      this.deps.finalizeMedia!,
      file,
      finalSize,
      this.deps.now(),
    );
  }

  private progress(fileId: number, bytes: number, expectedSize: number | null): void {
    setBytesDone(this.deps.db, fileId, bytes, this.deps.now());
    this.deps.emit("downloads://progress", { fileId, bytesDone: bytes, expectedSize });
  }

  private finish(fileId: number, end: TransferEnd): void {
    this.active.delete(fileId);
    applyEnd(this.deps.db, fileId, end, {
      nowMs: this.deps.now(),
      systemSuspended: this.systemSuspended,
      canTransfer: this.deps.canTransfer,
      retryDelaysMs: this.deps.retryDelaysMs ?? RETRY_DELAYS_MS,
    });
    this.notifyChanged();
    this.pump();
  }

  pause(fileId: number): void {
    pauseFile(this.deps.db, fileId, this.active.get(fileId), this.deps.now());
    this.notifyChanged();
  }

  resume(fileId: number): void {
    resumeFile(this.deps.db, fileId, this.deps.now());
    this.notifyChanged();
    this.pump();
  }

  /**
   * Relance les pauses SYSTÈME. Les pauses explicites ne bougent pas.
   *
   * Jusqu'ici, seul `start` les rattrapait — donc au prochain lancement de
   * l'application. Une machine endormie pendant la nuit rendait la main sur un
   * téléchargement à l'arrêt, sans que rien ne le reprenne. Branchée sur le
   * réveil de veille (`downloadsRuntime.ts`), cette méthode ferme ce trou.
   */
  resumeSystemPauses(): void {
    this.systemSuspended = false;
    if (this.creds === null) return;
    if (requeueSystemPauses(this.deps.db, this.deps.now()) === 0) return;
    this.notifyChanged();
    this.pump();
  }

  /**
   * Pause SYSTÈME de tout ce qui tourne ou attend — réseau cellulaire quand
   * seul le Wi-Fi est autorisé, par exemple. `paused_by_user` reste à 0 :
   * `resumeSystemPauses` relance tout au retour des conditions.
   */
  suspendForSystem(): void {
    this.systemSuspended = true;
    for (const flags of this.active.values()) flags.pause = true;
    const suspended = suspendQueued(this.deps.db, this.deps.now());
    if (suspended > 0 || this.active.size > 0) this.notifyChanged();
  }

  cancel(fileId: number): void {
    const flags = this.active.get(fileId);
    cancelFile(this.deps.db, this.deps.volume(), fileId, flags, this.deps.now());
    // Un transfert en vol posera son statut lui-même en se terminant.
    if (flags === undefined) this.notifyChanged();
  }

  /**
   * Attend la fin effective d'un transfert, pour que la suppression d'un claim
   * ne coure pas contre un worker qui écrit encore.
   */
  async waitNotActive(fileId: number, timeoutMs: number): Promise<void> {
    const limit = this.deps.now() + timeoutMs;
    while (this.isActive(fileId) && this.deps.now() < limit) {
      await new Promise((resolve) => setTimeout(resolve, 50));
    }
  }
}
